pallets/parachain-staking/src/nominator_list.rs

```diff
@@ -0,0 +1,151 @@
+	pub nominators: Vec<AccountId>,
+	/// All bonds, sorted by bonded amount
+	pub bonds: Vec<Bond<AccountId, Balance>>,
+	/// Maximum number of nominators that will be selected
+	pub max_selected: u32,
+	/// Amount of contribution from nominators that make the cut
+	pub contribution: Balance,
+	/// Total bond of all nominators (including those that don't make the cut)
+	pub total_bond_amount: Balance,
+}
+
+impl<
+	AccountId: Ord + Clone,
+	Balance: AtLeast32BitUnsigned + Ord + Copy + sp_std::ops::AddAssign + sp_std::ops::SubAssign,
+> CollatorNominators<AccountId, Balance>
+{
+	/// Insert a nominator. The nominator must not previously exist or will return an error.
+	pub fn insert_nominator(&mut self, account: &AccountId, bond: &Balance) -> Result<NominatorAdded<Balance>, ()> {
+		match self.nominators.binary_search(account) {
+			Ok(_) => Err(()),
+			Err(index) => {
+				self.nominators.insert(index, account.clone());
+				Ok(())
+			}
+		}?;
+
+		let index = self.find_bond_insertion_index(&bond);
+		let selected = self.is_selected(index);
+
+		// TODO: a helper (like is_selected()) would be nice if this pattern is consistent
+		// recalculate bond tallies
+		self.total_bond_amount.saturating_add(bond.clone());
+		if selected {
+			self.contribution.saturating_add(bond.clone());
+			// TODO: test carefully
+			if self.nominators.len() > self.max_selected as usize {
+				// we know we bumped someone out of the top here, so reduce by that amount
+				assert!(index + 1 < self.nominators.len(), "indexing logic error");
+				self.contribution.saturating_sub(self.bonds[index + 1].amount);
+			}
+		}
+
+		self.bonds.insert(index, Bond { owner: account.clone(), amount: bond.clone() });
+
+		if selected {
+			// TODO: does this match the expected value for AddedToTop?
+			Ok(NominatorAdded::AddedToTop { new_total: self.contribution, })
+		} else {
+			Ok(NominatorAdded::AddedToBottom)
+		}
+	}
+
+	/// remove a nominator. this is immediate; this struct doesn't concern itself with the need to
+	/// delay an exit, etc.
+	///
+	/// after removal, same accounting as insert_nominator()
+	pub fn remove_nominator(&mut self) {}
+
+	/// makes an adjustment (positive or negative) to some specific nominator's stake.
+	pub fn adjust_nominator_stake(&mut self) {}
+
+	/// adjust_num_selected_nominators. this adjusts the cutoff for "top" nominators.
+	///
+	/// if separate top vs. bottom containers for nominators are required in the design, it would
+	/// need to mutate those and ensure that they are sorted.
+	///
+	/// in any case, it needs to do some accounting to adjust the sum of top and bottom nominations.
+	pub fn adjust_num_selected_nominators(&mut self) {}
+
+	/// accessors
+	pub fn get_active_nominator_stake(&self) {} // get sum of top N nominator stakes
+	pub fn get_inactive_nominator_stake(&self) {} // get sum of not top N nominator stakes
+	pub fn get_total_nominator_stake(&self) {} // sum of the two above
```

Comment:
you plan to calculate this every time its called instead of storing each result as a field on the struct that is updated when it changes

Reply:
I was intending that these would simply return cached values and that the cost of those calculations are incurred whenever a value changes (much like is currently done).

Reply:
d63e6b4 should make this more clear